### PR TITLE
Protection agains closed context

### PR DESCRIPTION
**Pull request by polfeliu**
Hi, I've been investigating a crash (0xC0000005) on an application that makes use of pysoem. 

The application has several layers and threads and some logic wasn't right, and sometimes (depending on external conditions) ended up calling Network.send_receive_processdata() after having called Network.close(). This is clearly wrong from the application side, but not immediately obvious and was a bit hard to debug.

Network.close() calls ecx_closenic, that on windows 32 deletes the critical section mutexes and other structures that send_receive_processdata uses in C and when calling that on this context it crashes hard without any traceback.

I would rather have a controlled python crash instead of a hard one if something similar happens again, so here's an improvement to prevent more hard-crashes

### CODE REVIEW

**Comment by bnjmnp** [comment]
Thank you very much @polfeliu your contribution, makes totally sens to implement something like this.

File: src/pysoem/pysoem.pyx [hunk 1]
```diff
@@ -239,6 +241,13 @@ cdef class CdefMaster:
             ret_val = cpysoem.ecx_init_redundant(&self._ecx_contextt, &self._ecx_redport, ifname.encode('utf8'), ifname_red.encode('utf8'))
         if ret_val == 0:
             raise ConnectionError('could not open interface {}'.format(ifname))
+
+        self.context_initialized = True
+
+    def __check_context_is_initialized(self):
+        if not self.context_initialized:
+            raise EnvironmentError("Soem Context is not initialized or has been closed. Call Master.open() first")
```
Comment (bnjmnp) [inline src/pysoem/pysoem.pyx:249]: I'm not sure if the EnvironmentError exception type is the right fit. 
The python [docs](https://docs.python.org/3/library/exceptions.html#EnvironmentError) say: 
>The following exceptions are kept for compatibility with previous versions; starting from Python 3.3, they are aliases of [OSError](https://docs.python.org/3/library/exceptions.html#OSError).
>
>exception EnvironmentError

If something like this is run:
```py
with open("foo.txt") as f:
    pass

f.read()
```
Python throws an ValueError. (`ValueError: I/O operation on closed file.`)
Also not very fitting...
What about the [UnboundLocalError](https://docs.python.org/3/library/exceptions.html#UnboundLocalError) I already used at some places:
>exception UnboundLocalError
>
>    Raised when a reference is made to a local variable in a function or method, but no value has been bound to that variable. This is a subclass of [NameError](https://docs.python.org/3/library/exceptions.html#NameError).

File: src/pysoem/pysoem.pyx [hunk 2]
```diff
@@ -685,6 +712,8 @@ cdef class CdefSlave:
         """
         if self._ecx_contextt == NULL:
             raise UnboundLocalError()
```
Comment (bnjmnp) [inline src/pysoem/pysoem.pyx:714]: This two lines could actually be removed, right? As `__check_context_is_initialized` is doing the same?

File: src/pysoem/pysoem.pyx [hunk 3]
```diff
@@ -854,6 +893,8 @@ cdef class CdefSlave:
...
```
Comment (bnjmnp) [inline src/pysoem/pysoem.pyx:894]: Same here: This two lines could actually be removed, right? As `__check_context_is_initialized` is doing the same?

File: src/pysoem/pysoem.pyx [hunk 4]
```diff
@@ -877,6 +918,8 @@ cdef class CdefSlave:
...
```
Comment (bnjmnp) [inline src/pysoem/pysoem.pyx:919]: And here: This two lines could actually be removed, right? As `__check_context_is_initialized` is doing the same?

File: src/pysoem/pysoem.pyx (hunk 2 above)
Comment (polfeliu) [reply src/pysoem/pysoem.pyx:714]: I'm not sure it's doing the same.  context_initialized keeps track if the Master (Context) is initialized (called open and not called close).
The previous one checks if if the slave has a reference to it's c context. I could not find anywhere where this is unset, so I don't think it will be ever triggered, but it's a different error altogether.
The master could have been deleted, and the CDefSlave instance dereferenced from the master, but this is not done.

Comment (polfeliu) [reply src/pysoem/pysoem.pyx:714]: While checking this I've found a possible pitfall with the existing code.
CdefMaster can be dereferenced and then deleted by the python garbage collector.
This also deletes, or makes the memory invalid for _ecx_contextt, context_initialized, _ec_slave, _settings attributes of the Cython struct that are passed as a pointer to the slave with CdefMaster._get_slave(). 
So in theory an application could have a slave with non-null pointers that are invalid because the master was deleted.
I will try to replicate this and see if I can get a crash.

File: src/pysoem/pysoem.pyx (hunk 1 above)
Comment (polfeliu) [reply src/pysoem/pysoem.pyx:249]: You are right, the exception type is not that clear. I think the best choice is a custom exception, like the other exceptions that are thrown on pysoem

File: src/pysoem/pysoem.pyx (hunk 2 above)
Comment (polfeliu) [reply src/pysoem/pysoem.pyx:714]: Okay, I was able to replicate this but the crash is not as hard as I expected.

I created a test that replicates this:
```
import sys
from pysoem.pysoem import LargeObject
import pysoem
IFNAME = "\\Device\\NPF_{####################}" # Insert yours

def test():
    master = pysoem.Master()
    print(f"Reference count of master is {sys.getrefcount(master)}")
    # https://docs.python.org/3/library/sys.html#sys.getrefcount
    # 2 reference: variable and one temporal

    master.open(IFNAME)
    master.config_init()
    slave = master.slaves[0]

    for i in range(1000): # Check that we can do a lot of regular sdo reads
        slave.sdo_read(
            index=0x1018,
            subindex=0x01,
        )

    print(f"Reference count of master is {sys.getrefcount(master)}")
    # https://docs.python.org/3/library/sys.html#sys.getrefcount
    # Still 2 reference: variable and one temporal
    
    # Now "delete" the master. This could be any de-referencing via objects
    del master
    # Now it should have no references so it's deleted

    large_objects = []

    for i in range(1000):
        # Reading with the master destroyed. Unsafe
        # But somehow we still have an accessible slave that has references to the master that is destroyed
        slave.sdo_read(
            index=0x1018,
            subindex=0x01,
        )
        # But is successful for a while since the objects already still in the memory

        # We have to create objects that will take the memory
        large_objects.append(LargeObject())
        # And at the end something will break
```

And modified pysoem a bit to have more info:
Created the large object:
```
from libc.string cimport memset
cdef class LargeObject:
    cdef uint32_t container[0xb0]

    def __cinit__(self):
        memset(self.container, 0, sizeof(self.container))
        container_position = <uint32_t>&self.container
        size = sizeof(self.container)
        end = container_position + size

        r = range(container_position, end)
        print(f"container range {r}")
````

And added info to master:
```
CdefMaster
    .....

    def __del__(self):
        print("Master object is being destroyed")

    def __cinit__(self):
        first_element_position = <uint32_t>&self._ec_slave
        last_element_position = <uint32_t>&self.context_initialized

        cdef_master_position_range = range(first_element_position, last_element_position)
        print(f"cdef_master_position_range {cdef_master_position_range}")
```

The output is the following:
```
cdef_master_position_range range(3512352816, 3512523328)
Reference count of master is 2
Reference count of master is 2
Master object is being destroyed
container range range(3511733872, 3511734576)
container range range(3511726512, 3511727216)
container range range(3511729456, 3511730160)
container range range(3511735344, 3511736048)
container range range(3511734608, 3511735312)
container range range(3511723568, 3511724272)
container range range(3511725776, 3511726480)
container range range(3511738288, 3511738992)
container range range(3511727984, 3511728688)
container range range(3511730928, 3511731632)
container range range(3511724304, 3511725008)
container range range(3511728720, 3511729424)
container range range(3511730192, 3511730896)
container range range(3511727248, 3511727952)
container range range(3511732400, 3511733104)
container range range(3511722832, 3511723536)
container range range(3511733136, 3511733840)
container range range(3511725040, 3511725744)
container range range(3511410176, 3511410880)
container range range(3511403552, 3511404256)
container range range(3511406496, 3511407200)
container range range(3511407232, 3511407936)
container range range(3511411648, 3511412352)
container range range(3511399872, 3511400576)
container range range(3511407968, 3511408672)
container range range(3511412384, 3511413088)
container range range(3511404288, 3511404992)
container range range(3511413120, 3511413824)
container range range(3511410912, 3511411616)
container range range(3511408704, 3511409408)
container range range(3511400608, 3511401312)
container range range(3511402816, 3511403520)
container range range(3511405760, 3511406464)
container range range(3511413856, 3511414560)
container range range(3511414592, 3511415296)
container range range(3511399136, 3511399840)
container range range(3511405024, 3511405728)
container range range(3511401344, 3511402048)
container range range(3511409440, 3511410144)
container range range(3511402080, 3511402784)
container range range(3511417008, 3511417712)
container range range(3511421424, 3511422128)
container range range(3511417744, 3511418448)
container range range(3511423632, 3511424336)
container range range(3511424368, 3511425072)
container range range(3511428048, 3511428752)
container range range(3511419952, 3511420656)
container range range(3511425104, 3511425808)
container range range(3511425840, 3511426544)
container range range(3511430256, 3511430960)
container range range(3511426576, 3511427280)
container range range(3511420688, 3511421392)
container range range(3511422160, 3511422864)
container range range(3511418480, 3511419184)
container range range(3511427312, 3511428016)
container range range(3511419216, 3511419920)
container range range(3511428784, 3511429488)
container range range(3511422896, 3511423600)
container range range(3511429520, 3511430224)
container range range(3511430992, 3511431696)
container range range(3511415536, 3511416240)
container range range(3511416272, 3511416976)
container range range(3511433408, 3511434112)
container range range(3511440768, 3511441472)
container range range(3511439296, 3511440000)
container range range(3511444448, 3511445152)
container range range(3511441504, 3511442208)
container range range(3511437824, 3511438528)
container range range(3511445184, 3511445888)
container range range(3511434880, 3511435584)
container range range(3511434144, 3511434848)
container range range(3511437088, 3511437792)
container range range(3511440032, 3511440736)
container range range(3511445920, 3511446624)
container range range(3511432672, 3511433376)
container range range(3511442240, 3511442944)
container range range(3511435616, 3511436320)
container range range(3511442976, 3511443680)
container range range(3511438560, 3511439264)
container range range(3511446656, 3511447360)
container range range(3511443712, 3511444416)
container range range(3511447392, 3511448096)
container range range(3511431936, 3511432640)
container range range(3511436352, 3511437056)
container range range(3511462320, 3511463024)
container range range(3511448336, 3511449040)
container range range(3511457904, 3511458608)
container range range(3511451280, 3511451984)
container range range(3511454224, 3511454928)
container range range(3511449072, 3511449776)
container range range(3511463792, 3511464496)
container range range(3511456432, 3511457136)
container range range(3511452016, 3511452720)
container range range(3511458640, 3511459344)
container range range(3511463056, 3511463760)
container range range(3511455696, 3511456400)
container range range(3511452752, 3511453456)
container range range(3511450544, 3511451248)
container range range(3511454960, 3511455664)
container range range(3511459376, 3511460080)
container range range(3511453488, 3511454192)
container range range(3511460112, 3511460816)
container range range(3511449808, 3511450512)
container range range(3511457168, 3511457872)
container range range(3511460848, 3511461552)
container range range(3511461584, 3511462288)
container range range(3512358768, 3512359472)
container range range(3512352880, 3512353584)
FAILED
tests\context_test.py:9 (test)
def test():
        master = pysoem.Master()
    
        print(f"Reference count of master is {sys.getrefcount(master)}")
        # https://docs.python.org/3/library/sys.html#sys.getrefcount
        # 2 reference: variable and one temporal
    
        master.open(IFNAME)
    
        master.config_init()
    
        slave = master.slaves[0]
    
        for i in range(1000):
            slave.sdo_read(
                index=0x1018,
                subindex=0x01,
            )
    
        print(f"Reference count of master is {sys.getrefcount(master)}")
        # https://docs.python.org/3/library/sys.html#sys.getrefcount
        # Still 2 reference: variable and one temporal
        del master
        # Now it should have no references so it's deleted
    
        large_objects = []
    
        for i in range(1000):
            # Reading with the master destroyed. Unsafe
>           slave.sdo_read(
                index=0x1018,
                subindex=0x01,
            )

context_test.py:39: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   pysoem.pysoem.WkcError

src\pysoem\pysoem.pyx:777: WkcError
```

Checking the range of memory that the master did use and the ranges that the large object was using and it's clearly overwritting the previously freed memory, causing this crash.
image

I have no idea why is a WkcError, but I think is just coincidental. Every time I run the script different memory locations are used and the new objects could land anywhere...

In conclusion. It is essential that the slave keeps a python reference to the master so the python garbage collector does not destroy the master while there are slaves that are referenced to it. This will also unify the exception I created into a single one

File: src/pysoem/pysoem.pyx [hunk 5]
```diff
@@ -239,6 +241,13 @@ cdef class CdefMaster:
             ret_val = cpysoem.ecx_init_redundant(&self._ecx_contextt, &self._ecx_redport, ifname.encode('utf8'), ifname_red.encode('utf8'))
         if ret_val == 0:
             raise ConnectionError('could not open interface {}'.format(ifname))
+
+        self.context_initialized = True
+
+    def __check_context_is_initialized(self):
+        if not self.context_initialized:
+            raise UninitializedContextError("SOEM Context is not initialized or has been closed. Call Master.open() first")
```
Comment (bnjmnp) [inline src/pysoem/pysoem.pyx:249]: I'm still not fully convinced by the exception name. The concept fo the context is a SOEM internal concept that pysoem hides form the user. Because the exception type is part of API, the name should not be about the internal concept.

PySerial does a good job on his
```py
import serial

with serial.Serial("COM1") as s:
    pass

s.read()
```
PySerial throws this: `serial.serialutil.PortNotOpenError: Attempting to use a port that is not open`.

What about adopting this and naming the exception `InterfaceNotOpenError("..")` and the message `"Attempting to use an interface that is not open"`?
Instead of "Interface", other names could be "NetworkInterface" or "Adapter".

**Comment by bnjmnp** [comment]
Please check again your latest change (pass reference of master to the slave and reuse the method).
It doesn't seem to work for me.
If I run this:
```py
with pysoem.open(ifname) as master:
    master.config_init()
    master.slaves[0].sdo_read(0x1018, 1)
```
I get this error: `AttributeError: 'Master' object has no attribute '_CdefSlave__check_context_is_initialized'`.
I think the issue is that with this
```py
cdef class CdefSlave:
    ...
    cdef readonly CdefMaster _master
```
`_master` is declared a CdefMaster object not a reference to a CdefMaster.
I tried this:
```py
cdef class CdefSlave:
    ...
    cdef CdefMaster* _master
```
.. to make it a reference, but then the assignment `ethercat_slave._master = self` inside  `_get_slave()` fails to compile, and I don't know how to fix that.
Maybe for now it is easier to share only the reference to the `context_initialized` bool of the `CdefMaster`.
```py
cdef class CdefSlave:
    ...
    cdef cpysoem.boolean* _context_initialized
```
This also requires to remove the `readonly` for `CdefMaster`'s `context_initialized`, I think.
And of course you need add a `__check_context_is_initialized()` function for `CdefSlave`.

**Comment by polfeliu** [comment]
Something does not add up.

Are you sure you are building it? Something lingering on the cache?
I have no problem running this script (python 3.9):
`import pysoem

ifname = "XXX"

with pysoem.open(ifname) as master:
    master.config_init()
    r = master.slaves[0].sdo_read(0x1018, 1)
    print(r)
`

I have also enabled github actions on my fork an successfully re-tested the script again with the wheel created there:
https://github.com/polfeliu/pysoem/actions/runs/11102314587

The attribute is a python object so like everything python is a reference. See the pysoem.c that is generated, it is a pointer to master:
image
